File: iOS/Kalorie/Core/Models/MealTypeModel.swift
//
//  MealTypeDomain.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation
import MealKit

struct MealTypeDomain {

    // MARK: - Properties

    let id: String
    let name: String
    let startTime: Date
    let endTime: Date
}

extension [MealTypeDomain] {
    func mealType(at date: Date) -> MealTypeDomain? {
        guard let id = MealWindowsKt.mealWindowAt(minutes: date.minutesSinceMidnight, windows: mealWindows)?.id else {
            return nil
        }
        return first { $0.id == id }
    }

    func resolvedMealTypeId(for food: FoodConsumedDomain) -> String? {
        MealWindowsKt.resolvedMealWindowId(
            minutes: food.date.minutesSinceMidnight,
            pinnedId: food.mealTypeId,
            windows: mealWindows
        )
    }

    private var mealWindows: [MealWindow] {
        map {
            MealWindow(
                id: $0.id,
                startMinutes: $0.startTime.minutesSinceMidnight,
                endMinutes: $0.endTime.minutesSinceMidnight
            )
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/models/MealTypeModel.kt
package antoni.kalorie.core.models

import antoni.kalorie.core.utils.minutesSinceMidnight
import antoni.kalorie.mealkit.MealWindow
import antoni.kalorie.mealkit.mealWindowAt
import antoni.kalorie.mealkit.resolvedMealWindowId
import java.time.Instant
import java.time.ZoneId
import java.util.Locale

data class MealTypeDomain(
    val id: String,
    val name: String,
    val startMinutes: Int,
    val endMinutes: Int,
) {

    // MARK: - Functions

    companion object {
        fun clockTime(minutes: Int): String = String.format(Locale.ROOT, "%02d:%02d", minutes / 60, minutes % 60)
    }
}

fun List<MealTypeDomain>.mealType(date: Instant): MealTypeDomain? {
    val id = mealWindowAt(date.minutesSinceMidnight(), mealWindows())?.id ?: return null
    return firstOrNull { it.id == id }
}

fun List<MealTypeDomain>.resolvedMealTypeId(food: FoodConsumedDomain, zone: ZoneId = ZoneId.systemDefault()): String? =
    resolvedMealWindowId(food.date.minutesSinceMidnight(zone), food.mealTypeId, mealWindows())

private fun List<MealTypeDomain>.mealWindows(): List<MealWindow> =
    map {
        MealWindow(
            id = it.id,
            startMinutes = it.startMinutes,
            endMinutes = it.endMinutes,
        )
    }
